proposals/route-blinding.md

```diff
@@ -0,0 +1,308 @@
+* Ephemeral public keys are defined as: `E(i) = e(i) * G`.
+
+### Requirements
+
+A node `N(r)` wants to provide a blinded route `N(0) -> N(1) -> ... -> N(r)` that must be used
+to receive onion messages.
+
+* The channels used along that route may be either announced or unannounced.
+* When used for payments, intermediate nodes in the blinded route MUST NOT learn `payment_secret`.
+* Intermediate nodes in the blinded route MUST NOT learn the `node_id` or `scid` of other
+  intermediate nodes except for their immediate predecessor or successor.
+* Intermediate nodes in the blinded route MUST NOT learn their distance to the recipient `N(r)`.
+* Senders MUST NOT learn the real `node_id` and `scid` of the blinded intermediate hops after the
+  introduction point `N(0)`.
+* If `N(r)` creates multiple blinded routes to herself, senders MUST NOT be able to tell that these
+  routes lead to the same recipient (unless of course this information is leaked by higher layers
+  of the protocol, such as using the same `payment_hash`).
+
+### Encrypted data
+
+Route blinding introduces a new TLV field to the onion `tlv_payload`: the `encrypted_data`.
+
+This field is used to carry data coming from the builder of the route that cannot be modified by the
+sender. For route blinding it only needs to contain the `scid` to use when forwarding the message,
+but it may be extended with additional data in the future. It uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`encrypted_data`)
+2. data:
+    * [`...*byte`:`encrypted_data`]
+
+Once decrypted, the content of this encrypted payload is a TLV stream that contains information to
+identify the next node.
+
+### Creating a blinded route
+
+`N(r)` performs the following steps to create a blinded route:
+
+```text
+Initialization:
+
+  e(0) <- {0;1}^256
+  E(0) = e(0) * G
+
+Blinding:
+
+  For i = 0 to r-1:
+    ss(i) = H(e(i) * P(i)) = H(k(i) * E(i))         // shared secret known only by N(r) and N(i)
+    B(i) = HMAC256("blinded_node_id", ss(i)) * P(i) // Blinded node_id for N(i), private key known only by N(i)
+    rho(i) = HMAC256("rho", ss(i))                  // Key used to encrypt payload for N(i) by N(r)
+    e(i+1) = H(E(i) || ss(i)) * e(i)                // Ephemeral private key, only known by N(r)
+    E(i+1) = H(E(i) || ss(i)) * E(i)                // NB: N(i) must not learn e(i)
+
+Blinded route:
+  
+  (P(0),fees(0),cltv(0),encrypted_data(0))
+  (B(1),fees(1),cltv(1),encrypted_data(1))
+  ...
+  (B(r-1),fees(r-1),cltv(r-1),encrypted_data(r-1))
+```
+
+Note that this is exactly the same construction as Sphinx, but at each hop we use the shared secret
+to derive a blinded `node_id` for `N(i)` for which the private key will only be known by `N(i)`.
+
+The recipient needs to provide `E(0)` and the blinded route to potential senders.
+
+The `encrypted_data(i)` is encrypted with ChaCha20-Poly1305 using the `rho(i)` key, and
```

Comment:
the `rho(i)` key does only use the shared secret `rho(i) = HMAC256("rho", ss(i))` according to line 106.
Thus if we use `rho(i)` for encryption according to the text there is no `B(i) = HMAC256("blinded_node_id", ss(i)) * P(i)` being used in `rho(i)`.

Reply:
I don't understand your comment. Do you mean that `B(i)` doesn't depend on `rho(i)`? If that's what you mean yes, this is true, because I wanted to separate the blinding process (derivation of the `B(i)`) from the encryption of `encrypted_data`. Maybe it's unnecessary, but it doesn't do any harm, does it?